sway_libs/src/nft/extensions/burnable/burnable.sw

```diff
@@ -0,0 +1,50 @@
+library burnable;
+
+dep burnable_events;
+
+use burnable_events::BurnEvent;
+use ::nft::{errors::{AccessError, InputError}, nft_core::NFTCore, nft_storage::{BALANCES, TOKENS}};
+use std::{auth::msg_sender, hash::sha256, logging::log, storage::{get, store}};
+
+pub trait Burnable {
+    /// Deletes this token from storage and decrements the balance of the owner.
```

Comment:
Should we also decrement from MAX_SUPPLY?

Reply:
MAX_SUPPLY never changes, it is static. If there is a maximum supply of 10k, then when one burns a token the max supply is still 10k but only 9,999 exist. 

I've also implemented each extension so as to not interact with one. The burn, metadata, supply, and admin extensions are isolated.

However, where this does come into play for example is the burn functionality. It does remove the token from storage, but it does not remove the metadata. This would have to be done in a separate call implemented by the developer. 

Another example is in the mint function, there are no checks for the administrator or the supply. It will mint even if exceeding the max supply. If the check is not added by the developer using the library before calling the mint function, there is nothing preventing a new token from being minted. 

On one hand, this ensures that NFT libraries without these functionalities don't have additional gas overhead costs for checks that don't apply. On the other hand, devs now need to add these themselves. I've given them the tools but they need to use them properly.

I have been uncertain as to whether the modules should overlap or not, but have opted for the latter. To avoid the additional gas costs and ensure composability without running into any dependency problems down the road this is the approach I took. Thoughts? @matt-user @Braqzen

Reply:
I see that makes sense.  It's tough to say, but I think your implementation of the latter is probably the right way to go about it

Reply:
I would opt for the approach of letting the developer defensively code. I'd do it on a case-by-case basis to see what makes sense. I think it's better to have a barebones implementation and allow the dev to build in the safety i.e. you can mint an unlimited about but the dev needs to ensure that you don't. It's tricky, without writing the whole thing myself and considering all of the work in unison I cannot say much more because "it depends".